c_tests/tests/noopts.c

```diff
@@ -0,0 +1,50 @@
+// Compiler:
+// Run-time:
+//   env-var: YKD_PRINT_IR=1
+//   stderr:
+//     ...
+//     store i32 2...
+//     ...
+//     store i32 3...
+//     ...
+//     ...add...
+//     ...
+
+// Check that our NOOPT macro blocks optimisations.
+
+#include <assert.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <yk_testing.h>
+
+int add_them(int x, int y) { return x + y; }
+
+int main(int argc, char **argv) {
+  int x, y, res;
+  void *tt = __yktrace_start_tracing(HW_TRACING, &res, &x, &y);
+  // The __yktrace_start_tracing() call has almost certainly already blocked
```

Comment:
This paragraph isn't necessarily true *if* `__yktrace_start_tracing` can be inlined. I'm not sure the comment is very helpful TBH.

Reply:
`__yktrace_start_tracing()` is in an external shared object, so it can't be inlined.

The reason I included this comment is because the first version of the test initialised `x` and `y` outside of the traced section of code and the compiler was already unable (without `NOOPT`) to replace the call with a constant due to `__yktrace_start_tracing`.

Reply:
OK. I think I'd just remove the comment to be honest.

Reply:
I don't often push back, but I think this is a critical comment.

Someone may be tempted to move the initialisation up outside of the traced section, at which point we no longer test `NOOPT`. At that point we'd be testing if `__yk_start_tracing()` blocks optimisations, which is not the intention of the test.

Reply:
Surely all I have to do is something like:

```c
int x = 1;
start_tracing(&x);
x = 2;
NOOPT(x);
...
```

? In other words, I don't think *initialisation* is very special per se.

Reply:
OK, you got me.

s/initialisation/assignment/

We can't expect the compiler to constant fold an *assignment* that happened before the call to `start_tracing()`, as it doesn't know if it was mutated.

What I don't want is for someone to change the test to the (shorter):
```
int x = 1, y = 2;
NOOPT(X);
NOOPT(Y);
start_tracing()
...
```

As here, whether or not `NOOPT` blocks optimisations is moot, as we know that (currently) `start_tracing` already blocks optimisations. Thus we wouldn't be testing what we thought we were. `NOOPT` could be `#define NOOPT(x) x` and the test would still pass.

Does that make sense?

Reply:
I now realise that (once again!) there are two levels of IR we're thinking about here: `NOOPT` is meant to inhibit optimisations in both the AOT and JIT IRs. [I must get used to thinking at both levels!] I think our test is meant to test only that it does so in the JIT IR? If that is the case (and it's probably worth being clear about that in a comment), then it's obvious that the "IR part of the test" can only happen between stop/start tracing.

Reply:
I think here we'd be checking that it doesn't happen in the AOTMod, since `YKD_PRINT_IR` prints the `IR` before it's submitted to the execution engine.

TBH, I don't care where we test if `NOOPT` works, I just want to know that it does work!

Reply:
In our case I think we do want to know that it works at both IR levels! It would be confusing if it worked in the AOT IR but not the JIT IR. It wouldn't hurt to test both IRs, but at the moment we only have the JIT IR debugging as output -- so we might as well test that. In fairness, if it works in the JIT IR, it almost certainly works in the AOT IR!

Reply:
OK. Testing the post-opt JIT module would require a larger change (and an additional test). 

Let's put this on hold until @ptersilie can figure out the blockmap bug.